ProcessLib/ThermoMechanicalPhaseField/ThermoMechanicalPhaseFieldFEM-impl.h

```diff
@@ -0,0 +1,328 @@
+    assembleWithJacobianForHeatConductionEquations(
+        double const t, std::vector<double> const& local_xdot,
+        const double /*dxdot_dx*/, const double /*dx_dx*/,
+        std::vector<double>& /*local_M_data*/, std::vector<double>& /*local_K_data*/,
+        std::vector<double>& local_b_data, std::vector<double>& local_Jac_data,
+        LocalCoupledSolutions const& local_coupled_solutions)
+{
+    auto const& local_d =
+        local_coupled_solutions.local_coupled_xs[_phase_field_process_id];
+    auto const& local_T =
+        local_coupled_solutions.local_coupled_xs[_heat_conduction_process_id];
+    assert(local_T.size() == temperature_size);
+    assert(local_d.size() == phasefield_size);
+
+    auto d = Eigen::Map<
+        typename ShapeMatricesType::template VectorType<phasefield_size> const>(
+        local_d.data(), phasefield_size);
+
+    auto T = Eigen::Map<typename ShapeMatricesType::template VectorType<
+        temperature_size> const>(local_T.data(), temperature_size);
+
+    auto T_dot = Eigen::Map<typename ShapeMatricesType::template VectorType<
+        temperature_size> const>(local_xdot.data(), temperature_size);
+
+    auto local_Jac = MathLib::createZeroedMatrix<
+        typename ShapeMatricesType::template MatrixType<
+            temperature_size, temperature_size>>(
+        local_Jac_data, temperature_size, temperature_size);
+
+    auto local_rhs =
+        MathLib::createZeroedVector<typename ShapeMatricesType::
+                                        template VectorType<temperature_size>>(
+            local_b_data, temperature_size);
+
+    double const& dt = _process_data.dt;
+
+    SpatialPosition x_position;
+    x_position.setElementID(_element.getID());
+
+    int const n_integration_points = _integration_method.getNumberOfPoints();
+    for (int ip = 0; ip < n_integration_points; ip++)
+    {
+        x_position.setIntegrationPoint(ip);
+        auto const& w = _ip_data[ip].integration_weight;
+        auto const& N = _ip_data[ip].N;
+        auto const& dNdx = _ip_data[ip].dNdx;
+
+        auto& eps_m = _ip_data[ip].eps_m;
+        auto& heatflux = _ip_data[ip].heatflux;
+
+        auto rho_sr = _process_data.solid_density(t, x_position)[0];
+        auto const alpha = _process_data.linear_thermal_expansion_coefficient(
+            t, x_position)[0];
+        double const c = _process_data.specific_heat_capacity(t, x_position)[0];
+        auto const lambda =
+            _process_data.thermal_conductivity(t, x_position)[0];
+        auto const lambda_res =
+            _process_data.residual_thermal_conductivity(t, x_position)[0];
+        double const T0 = _process_data.reference_temperature;
+
+        double const d_ip = N.dot(d);
+        double const T_ip = N.dot(T);
+        double const T_dot_ip = N.dot(T_dot);
+        double const delta_T = T_ip - T0;
+        // calculate real density
+        double const rho_s = rho_sr * (1 - 3 * alpha * delta_T);
```

Comment:
Correct it or use rho_sr as rho_s directly. For solid of geomaterials or metal, the factor 1 / (1 +3 * alpha * dT) to their density are tiny.